src/interface/variable.cpp

```diff
@@ -85,28 +86,42 @@ CellVariable<T>::AllocateCopy(const bool allocComms, std::weak_ptr<MeshBlock> wp
 /// Initialize a 6D variable
 template <typename T>
 void CellVariable<T>::allocateComms(std::weak_ptr<MeshBlock> wpmb) {
-  // set up fluxes
+  if (wpmb.expired()) return;
+  std::shared_ptr<MeshBlock> pmb = wpmb.lock();
   std::string base_name = label();
+
+  // TODO(JMM): Note that this approach assumes LayoutRight. Otherwise
+  // the stride will mess up the types.
+
+  // Compute size of unified comm_data object and create it. A unified
+  // comm_data_ object reduces the number of memory allocations per
+  // variable per meshblock from 5 to 2.
+  int n_outer = 0;
   if (IsSet(Metadata::Independent)) {
-    flux[X1DIR] = ParArrayND<T>(base_name + ".fluxX1", GetDim(6), GetDim(5), GetDim(4),
-                                GetDim(3), GetDim(2), GetDim(1));
-    if (GetDim(2) > 1)
-      flux[X2DIR] = ParArrayND<T>(base_name + ".fluxX2", GetDim(6), GetDim(5), GetDim(4),
-                                  GetDim(3), GetDim(2), GetDim(1));
-    if (GetDim(3) > 1)
-      flux[X3DIR] = ParArrayND<T>(base_name + ".fluxX3", GetDim(6), GetDim(5), GetDim(4),
-                                  GetDim(3), GetDim(2), GetDim(1));
+    n_outer += (GetDim(1) > 1) + (GetDim(2) > 1) + (GetDim(3) > 1);
   }
+  n_outer += (pmb->pmy_mesh->multilevel);
+  comm_data_ = ParArray7D<T>(base_name + ".comm_data", n_outer, GetDim(6), GetDim(5),
```

Comment:
Is `comm_data_` only the collection of 4 arrays - the 3 fluxes and the coarse array? Does the benefit of combining these allocations outweight the cost of wasted memory for `coarse_s`? Could `coarse_s` remain it's own array?

Reply:
This is less relevant for this PR, but is allocating space for all the fluxes necessary? Could the flux divergence be computed in each dimension one by one so that only one flux array is needed in memory? If possible, this would allow larger problem sizes but might be too big a code change.

Reply:
> Is comm_data_ only the collection of 4 arrays - the 3 fluxes and the coarse array? Does the benefit of combining these allocations outweight the cost of wasted memory for coarse_s? Could coarse_s remain it's own array?

That could be done. I'm not sure what the right trade-off is. I would say we accept the memory overhead for now, and we can split out the coarse buffer later if we desire.

> This is less relevant for this PR, but is allocating space for all the fluxes necessary? Could the flux divergence be computed in each dimension one by one so that only one flux array is needed in memory? If possible, this would allow larger problem sizes but might be too big a code change.

This is an interesting idea. I agree not for this PR, but something to think about.